drivers/wifi/siwx91x/siwx91x_wifi.c

```diff
@@ -0,0 +1,368 @@
+
+	memcpy(tmp.ssid, result->scan_info[item].ssid, tmp.ssid_length);
+	memcpy(tmp.mac, result->scan_info[item].bssid, tmp.mac_length);
+	for (i = 0; i < ARRAY_SIZE(security_convert); i++) {
+		if (security_convert[i].sl_val == result->scan_info[item].security_mode) {
+			tmp.security = security_convert[i].z_val;
+		}
+	}
+	sidev->scan_res_cb(sidev->iface, 0, &tmp);
+}
+
+static unsigned int siwx91x_on_scan(sl_wifi_event_t event, sl_wifi_scan_result_t *result,
+				    uint32_t result_size, void *arg)
+{
+	struct siwx91x_dev *sidev = arg;
+	int i;
+
+	if (!sidev->scan_res_cb) {
+		return -EFAULT;
+	}
+	for (i = 0; i < result->scan_count; i++) {
+		siwx91x_report_scan_res(sidev, result, i);
+	}
+	sidev->scan_res_cb(sidev->iface, 0, NULL);
+	sidev->state = WIFI_STATE_INACTIVE;
+	return 0;
+}
+
+static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_scan_config,
+			scan_result_cb_t cb)
+{
+	sl_wifi_scan_configuration_t sl_scan_config = { };
+	struct siwx91x_dev *sidev = dev->data;
+	int ret;
+
+	__ASSERT(z_scan_config, "z_scan_config cannot be NULL");
+
+	if (sidev->state != WIFI_STATE_INACTIVE) {
+		return -EBUSY;
+	}
+
+	/* The enum values are same, no conversion needed */
+	sl_scan_config.type = z_scan_config->scan_type;
+
+	sl_scan_config.channel_bitmap_2g4 = 0xFFFF;
+	memset(sl_scan_config.channel_bitmap_5g, 0xFF, sizeof(sl_scan_config.channel_bitmap_5g));
+
+	sidev->scan_res_cb = cb;
+	ret = sl_wifi_start_scan(SL_WIFI_CLIENT_INTERFACE, NULL, &sl_scan_config);
+	if (ret != SL_STATUS_IN_PROGRESS) {
+		return -EIO;
+	}
+	sidev->state = WIFI_STATE_SCANNING;
+
+	return 0;
+}
+
+static int siwx91x_status(const struct device *dev, struct wifi_iface_status *status)
+{
+	struct siwx91x_dev *sidev = dev->data;
+	int32_t rssi = -1;
+
+	memset(status, 0, sizeof(*status));
+	status->state = sidev->state;
+	sl_wifi_get_signal_strength(SL_WIFI_CLIENT_INTERFACE, &rssi);
+	status->rssi = rssi;
+	return 0;
+}
+
+#ifdef CONFIG_WIFI_SILABS_SIWX91X_NET_STACK_NATIVE
+
+static int siwx91x_send(const struct device *dev, struct net_pkt *pkt)
+{
+	size_t pkt_len = net_pkt_get_len(pkt);
+	struct net_buf *buf = NULL;
+	int ret;
+
+	if (net_pkt_get_len(pkt) > _NET_ETH_MAX_FRAME_SIZE) {
+		LOG_ERR("unexpected buffer size");
+		return -ENOBUFS;
+	}
+	buf = net_buf_alloc(&siwx91x_tx_pool, K_FOREVER);
+	if (!buf) {
+		return -ENOBUFS;
+	}
+	if (net_pkt_read(pkt, buf->data, pkt_len)) {
+		net_buf_unref(buf);
+		return -ENOBUFS;
+	}
+	net_buf_add(buf, pkt_len);
+
+	ret = sl_wifi_send_raw_data_frame(SL_WIFI_CLIENT_INTERFACE, buf->data, pkt_len);
+	if (ret) {
+		return -EIO;
+	}
+
+	net_pkt_unref(pkt);
+	net_buf_unref(buf);
+
+	return 0;
+}
+
+/* Receive callback. Keep the name as it is declared weak in WiseConnect */
+sl_status_t sl_si91x_host_process_data_frame(sl_wifi_interface_t interface,
+					     sl_wifi_buffer_t *buffer)
+{
+	sl_si91x_packet_t *si_pkt = sl_si91x_host_get_buffer_data(buffer, 0, NULL);
+	struct net_if *iface = net_if_get_default();
```

Comment:
You might get wrong interface if there are multiple interfaces in the system. I suggest using `net_if_get_first_wifi();` as I suppose you want to get a wifi interface here.

Reply:
Also fixed the occurrences in `siwx91x_wifi_socket.c`.